File: Sources/Helpers/DotFile.swift
//
//  DotFile.swift
//  
//
//  Created by Owen Cruz-Abrams on 9/23/23.
//

import Foundation

//TODO: Should we integrate this into logic for Init, Input, etc?
struct DotFile: Codable {
    var location: URL
    
    var aliases: [DotFile.Alias]
    
    init(location: URL, aliases: [DotFile.Alias]) {
        self.location = location
        self.aliases = aliases
    }
    
    
    /// Print the ``DotFile`` to STDOUT.
    /// - Parameter docs: Whether to print out the docs for each alias in addition to the alias itself.
    func output(docs: Bool) {
        print("You have ", terminator: "")
        boldPrint("\(self.aliases.count)", terminator: " ")
        print("aliases added in Ally.")
        for i in self.aliases {
            if docs {
                if let idocs = i.docs {
                    boldPrint(idocs)
                }
                
            }
            boldPrint("\(i.aliasName)", terminator: " -> ")
            print(i.longFormCommand)
        }
    }
}

//MARK: - Parsing extensions + types
extension DotFile {
    struct Alias: Codable {
        var docs: String?
        var aliasName: String
        var longFormCommand: String
        
        init(docs: String? = nil, aliasName: String, longFormCommand: String) {
            self.docs = docs
            self.aliasName = aliasName
            self.longFormCommand = longFormCommand
        }
    }
    
    static func parseAllyFile(fileLocation: URL) throws -> DotFile {
        
        func parseAlreadyDocs(alreadyDocs: String, index: Int, lines: [String]) -> [Self.Alias] {
            var returnable = [Self.Alias]()
            if index > lines.count-1 {
                return returnable
            }
            let currentLine = lines[index]
            if currentLine.trimmingCharacters(in: .whitespacesAndNewlines).starts(with: "#") {
                let alDocs = alreadyDocs + "\n" + currentLine
                returnable += parseAlreadyDocs(alreadyDocs: alDocs, index: index+1, lines: lines)
            } else {
                let parts = currentLine.split(separator: "=").map { sub in
                    String(sub)
                }
                var firstPart = parts[0]
                let firstPartFifthIndex = firstPart.index(firstPart.startIndex, offsetBy: 5)
                firstPart = String(firstPart[firstPartFifthIndex...])
                firstPart = firstPart.replacingOccurrences(of: "alias ", with: "").trimmingCharacters(in: .whitespacesAndNewlines)
                let secondPart = parts[1]
                returnable.append(Self.Alias.init(docs: alreadyDocs, aliasName: firstPart, longFormCommand: secondPart))
                returnable += parseAlreadyDocs(alreadyDocs: "", index: index+1, lines: lines)
                    
                
            }
            
            return returnable
        }
        // Get the contents
        let fileContents = try String(contentsOf: fileLocation, encoding: .utf8)
        // Split by newlines. While aliases will only take up one line, it's worth noting that docs can take up several. Thus, we parse by the end of each line NOT containing "#"
        let fileContentsList = fileContents.split(separator: "\n").map { sub in
            String(sub)
        }
        let aliases: [Self.Alias] = parseAlreadyDocs(alreadyDocs: "", index: 0, lines: fileContentsList)
        
        return DotFile(location: fileLocation, aliases: aliases)
        
        
    }
}

File: src/helpers/dot_file.py
import os

class DotFile:
    def __init__(self, location):
        self.location = location
        self.aliases = []

    def output(self, docs):
        print("You have ", end="")
        self.bold_print(f"{len(self.aliases)}", end=" ")
        print("aliases added in Ally.")
        for alias in self.aliases:
            if docs and alias.docs:
                self.bold_print(alias.docs)
            self.bold_print(f"{alias.alias_name}", end=" -> ")
            print(alias.long_form_command)

    @staticmethod
    def bold_print(text, end="\n"):
        print(f"\033[1m{text}\033[0m", end=end)

    class Alias:
        def __init__(self, docs=None, alias_name="", long_form_command=""):
            self.docs = docs
            self.alias_name = alias_name
            self.long_form_command = long_form_command

    @staticmethod
    def parse_ally_file(file_location):
        def parse_already_docs(already_docs, index, lines):
            returnable = []
            if index > len(lines) - 1:
                return returnable
            current_line = lines[index]
            if current_line.strip().startswith("#"):
                al_docs = already_docs + "\n" + current_line
                returnable += parse_already_docs(al_docs, index + 1, lines)
            else:
                parts = current_line.split("=")
                first_part = parts[0].replace("alias ", "").strip()
                second_part = parts[1]
                returnable.append(DotFile.Alias(already_docs, first_part, second_part))
                returnable += parse_already_docs("", index + 1, lines)
            return returnable

        with open(file_location, "r", encoding="utf-8") as file:
            file_contents = file.read()

        file_contents_list = file_contents.split("\n")
        aliases = parse_already_docs("", 0, file_contents_list)

        dot_file = DotFile(file_location)
        dot_file.aliases = aliases
        return dot_file
